### DISCUSSION TITLE
Effects for A2DP and flash size question

### DISCUSSION THREAD

**Discussion by CursingThomas** [pschatzmann/arduino-audio-tools#1197] [Q&A] [answered]
Hi all, 

I have a combined question!
I was digging around and wanted to implement something which allows an A2DP input to be outputted to I2s (my external DAC). I did this succesfully! And it works flawlessly. 

However i wanted to update this, by enabling myself to put effects on this. Though, when i thought i had found the solution (A2DP stream in combination with a copier) my esp32 told me i was using too much flash memory, and i could not upload my sketch.

Therefore i have two questions:
1. Is there a more memory efficient way to still enable effects or filters onto a a2dp->i2s setup?
2. What microcontrollers do you guys use? I came from a subproject of the same author (bluetootha2dpsink) and there the esp32 was strongly mentioned. What microcontroller was used to make the examples?

Thanks in advance!

Edit:

I have solved my memory issue, but still am stuck on the effects problem. I cannot seem to find examples or work from other members which showcases how to use effects on bluetooth connections. Can someone help me out with this? I would like to add a bit of gain to a rock song for example, and listen to it via aux (i2s)

**Comment by pschatzmann** [comment, marked as answer]
How about reading the the documentation: you can find the answer in the [A2DPStream](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_a2_d_p_stream.html).

The ESP32 usually has enough flash size: you just need to select the Partition Scheme that provides enough memory for APP

**Reply by CursingThomas** [reply]
Hi, i have read that file. Alongside having tried many examples. Though i am still not succesfull in getting effects on my stream.

Currently, the main logic of my code is:

A2DPStream input;
I2SStream output;
Streamcopy copy(output, x);

This works wonderfull, but again; i have no clue how i should go about adding effects into this configuration.

**Reply by pschatzmann** [reply]
I think you missed the point: the recommendation is not to use A2DPStream!

**Reply by CursingThomas** [reply]
AH, like that; then yes, I missed your point. My apologies.
Do you have examples available on how to use effects with an a2dp sink instead? What is the way to go about this. I read in another Q&A about using a callback, though I have no clue on how this should be setup.  

If you could provide me with more information that would be highly appreciated.

**Reply by pschatzmann** [reply]
As the documentation tells you, the examples can be found in basic-api: e.g. [this](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/basic-a2dp-i2s/basic-a2dp-i2s.ino). 

In the callback you just write to an AudioEffectStream object.

**Reply by CursingThomas** [reply]
I used the example you provided, and added this:
BluetoothA2DPSink a2dp_sink;
I2SStream i2s;
AudioEffectStream effects(i2s);
ADSRGain adsr(0.0001,0.0001, 0.9 , 0.0002);
Fuzz fuzz(20, 300);

// Write data to I2S
void read_data_stream(const uint8_t *data, uint32_t length) {
  i2s.write(data, length);
  effects.write(data, length);
}

Then in the setup i use effect.jaddEffect(fuzz);

Is this the correct way to write to the object? I am not getting any change on my output. Thanks for the help so far.

**Comment by pschatzmann** [comment]
Just look at the [Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/Audio-Effects). I think you forgot to call begin

**Reply by CursingThomas** [reply]
I did not sadly, I adjusted my code to be compliant with the example and now i get a very garbled slowed down output which has nothing to do with fuzz. 

My sketch code:
`#include "BluetoothA2DPSink.h"
#include "AudioTools.h"

BluetoothA2DPSink a2dp_sink;
I2SStream i2s;
AudioEffectStream effects(i2s);
ADSRGain adsr(0.0001,0.0001, 0.9 , 0.0002);
Fuzz fuzz(1, 300);
StreamCopy copier (i2s, effects);

// Write data to I2S
void read_data_stream(const uint8_t *data, uint32_t length) {
  i2s.write(data, length);
  effects.write(data, length);
}

void setup() {
  Serial.begin(115200);

  // register callback
  a2dp_sink.set_stream_reader(read_data_stream, false);
  
  // Start Bluetooth Audio Receiver
  a2dp_sink.set_auto_reconnect(false);
  a2dp_sink.start("a2dp-i2s");
  
  effects.addEffect(fuzz);
  // setup output
  auto cfg = i2s.defaultConfig();
  cfg.pin_bck = 26;
  cfg.pin_ws = 25;
  cfg.pin_data = 22;
  i2s.begin(cfg);
  effects.begin(cfg);
  
}

void loop() { 
  copier.copy();
 }`
 
 Any idea?

**Comment by pschatzmann** [comment]
Your callback is wrong: why are you writing to i2s ?

**Reply by CursingThomas** [reply]
I thought it was a generic callback which i could use, and that since it was my output i should write to that. Can you provide an example on how to properly do this? Since i am thoroughly lost right now.

**Reply by pschatzmann** [reply]
Just delete the line: i2s.write(data, length);
The effects.write is alreday sending it's output to i2s!

**Reply by CursingThomas** [reply]
You are right, that does make a lot of sense now that i think about it.
My problems are fixed, massive thanks.

**Reply by CursingThomas** [reply]
hi, Sadly i have other problems i am encountering.
First of all, it seems like some effects are barely if not wroking at all. The distortion for example doesnt do that much. 
Besides, the application crashes after 3-4 seconds or so. This is my code: 
#include "BluetoothA2DPSink.h"
#include "AudioTools.h"

BluetoothA2DPSink a2dp_sink;
I2SStream i2s;
AudioEffectStream effects(i2s);
ADSRGain adsr(0.0001,0.0001, 0.9 , 0.0002);
Fuzz fuzz(10, 300);
StreamCopy copier (i2s, effects);
Distortion diss(9000, 10000);
PitchShift shifter (2, 1000);

// Write data to I2S
void read_data_stream(const uint8_t *data, uint32_t length) {
  effects.write(data, length);
}

void setup() {
  Serial.begin(115200);

  // register callback
  a2dp_sink.set_stream_reader(read_data_stream, false);
  
  // Start Bluetooth Audio Receiver
  a2dp_sink.set_auto_reconnect(false);
  a2dp_sink.start("a2dp-i2s");
  
  // setup output
  auto cfg = i2s.defaultConfig();
  cfg.pin_bck = 26;
  cfg.pin_ws = 25;
  cfg.pin_data = 22;
  i2s.begin(cfg);
  effects.begin(cfg);
  effects.addEffect(shifter);
  
}

void loop() { 
  copier.copy();
 }

**Comment by pschatzmann** [comment]
If it does not do much, then it is because of your selected parameters...

**Reply by CursingThomas** [reply]
Okay, i thought i had them pretty well covered. Anyways, the pitch shift in the above example works fine. Though it crashes like i mentioned after a few seconds. Any ideas?
